### PR TITLE
glsl.meta.slang image #3631, Review

### PR DESCRIPTION
**Pull request by ArielG-NV**

### CODE REVIEW

File: slang-gfx.h [hunk 1]
```diff
@@ -2724,5 +2732,4 @@ struct SlangSessionExtendedDesc
     uint32_t compilerOptionEntryCount = 0;
     slang::CompilerOptionEntry* compilerOptionEntries = nullptr;
 };
-
```
Comment (jkwak-work) [inline slang-gfx.h]: Please don't remove lines unnecessarily.

File: source/slang/glsl.meta.slang [hunk 2]
```diff
@@ -2821,6 +2821,453 @@ public vec4 shadow2DProjLod(sampler2DShadow sampler, vec4 coord, float lod)
...
+
+public const int GL_TEXTURE_CUBE_MAP_POSITIVE_X = 0;
+public const int GL_TEXTURE_CUBE_MAP_NEGATIVE_X = 1;
+public const int GL_TEXTURE_CUBE_MAP_POSITIVE_Y = 2;
+public const int GL_TEXTURE_CUBE_MAP_NEGATIVE_Y = 3;
+public const int GL_TEXTURE_CUBE_MAP_POSITIVE_Z = 4;
+public const int GL_TEXTURE_CUBE_MAP_NEGATIVE_Z = 5;
+
+__generic<T : __BuiltinType>
+[ForceInline] void typeRequireChecks_image_atomic_teir1()
```
Comment (jkwak-work) [inline source/slang/glsl.meta.slang]: Did you mean "tier"?

File: source/slang/slang-check-decl.cpp [hunk 3]
```diff
@@ -944,7 +944,27 @@ namespace Slang
...
+                    {
+                        isLValue = false;
+                        qualType.hasReadOnlyOnTarget = true;
+                        if (isLValue == false && isWriteOnly) break;
+                    }
+                    if (as<GLSLWriteOnlyModifier>(mod))
+                    {
+                        isWriteOnly = true;
+                        if (isLValue == false && isWriteOnly) break;
+                    }
```
Comment (jkwak-work) [inline source/slang/slang-check-decl.cpp:962]: Does this mean it is legal to have both "readonly" and "writeonly"?
Or is it an error case to have both of modifiers?

File: tools/render-test/shader-input-layout.cpp [hunk 4]
```diff
@@ -1100,12 +1126,82 @@ namespace renderer_test
...
+                break;
+            }
+            }
+        }
+    }
     void generateTextureData(TextureData& output, const InputTextureDesc& desc)
     {
         gfx::FormatInfo formatInfo;
         gfxGetFormatInfo(desc.format, &formatInfo);
-
```
Comment (jkwak-work) [inline tools/render-test/shader-input-layout.cpp]: I like to discourage to from making an unnecessary line removal.

File: tools/render-test/shader-input-layout.cpp [hunk 5]
```diff
@@ -1139,12 +1235,12 @@ namespace renderer_test
                     const uint8_t* srcPixels = (const uint8_t*)srcSlice.values;
 
                     int16_t* dstPixels = (int16_t*)output.setSliceCount(i, pixelCount);
-
```
Comment (jkwak-work) [inline tools/render-test/shader-input-layout.cpp:1142]: I like to discourage to from making an unnecessary change.

File: tools/render-test/shader-input-layout.cpp [hunk 6]
```diff
@@ -1139,12 +1235,12 @@ namespace renderer_test
                     const uint8_t* srcPixels = (const uint8_t*)srcSlice.values;
 
                     int16_t* dstPixels = (int16_t*)output.setSliceCount(i, pixelCount);
-                    
+
                     switch (formatInfo.channelCount)
                     {
                         case 1:
                         {
-                            for (Index j = 0; j < pixelCount; ++j, srcPixels += 4, dstPixels += 1)
```
Comment (jkwak-work) [inline tools/render-test/shader-input-layout.cpp:1147]: I like to discourage to from making an unnecessary change.

File: tools/render-test/shader-input-layout.cpp [hunk 7]
```diff
@@ -1100,12 +1126,82 @@ namespace renderer_test
...
 
+            T* dstPixels = (T*)output.setSliceCount(i, pixelCount);
+            switch (formatInfo.channelCount)
+            {
+            case 1:
+            {
+                for (Index j = 0; j < pixelCount; ++j, srcPixels += 4, dstPixels++)
+                {
+                    // Copy out r
+                    dstPixels[0] = T(srcPixels[0] * (1.0f / 255));
```
Comment (jkwak-work) [inline tools/render-test/shader-input-layout.cpp]: I am not sure if this is a right way for integer types.
The original code looks like the assumption is that srcPixels is 8 bit format whose value was normalized.
And it is converting to a normalized floating value.
As an example, when srcPixels[0] was 254, dstPixels[0] would get 0.99f as a floating value, which makes sense.
But if the destination type is an integertype and srcPixels[0] is 254, dstPixels[0] will get 0 because int(254 / 255) is zero. The resulting value will be always zero or one, which doesn't make sense.

I think that when the destination is an integer value, you need to multiply the result with the max value of the type to keep it normalized.
Something like this,
```
dstPixels[0] = T(srcPixels[0] * limits<T>::max() / 255);
```

File: tools/gfx-unit-test/format-unit-tests.cpp [hunk 8]
```diff
@@ -337,6 +337,22 @@ namespace gfx_test
...
+
+            auto texView = createTexView(device, size, gfx::Format::R64_UINT, &subData);
+            setUpAndRunTest(device, texView, uintBufferView, "copyTexUint5");
+            compareComputeResult(
+                device,
+                uintResults,
+                Slang::makeArray<uint32_t>(255u, 0u, 0u, 255u,
+                                           255u, 255u, 127u, 127u));
+        }
+
```
Comment (jkwak-work) [inline tools/gfx-unit-test/format-unit-tests.cpp]: You will need to either remove this or make it working.

**Review by jkwak-work — COMMENTED** [review]
I have a few comments.
I haven't finished reviewing yet.
I will continue tomorrow.

File: source/slang/slang-check-decl.cpp (hunk 3 above)
Comment (ArielG-NV) [reply source/slang/slang-check-decl.cpp:962]: it is legal to have both, it is legal to compound any number of memory quaifier's

File: tools/gfx-unit-test/format-unit-tests.cpp (hunk 8 above)
Comment (ArielG-NV) [reply tools/gfx-unit-test/format-unit-tests.cpp]: I can't make the test working because the test uses a type that does not support uniform (non storage) use

the test is here for the same reason `auto texView = createTexView(device, size, gfx::Format::R32G32B32_FLOAT, &subData);` is added, neither are allowed to work with the resources in the test, but was added for completion(?)

File: tools/render-test/shader-input-layout.cpp (hunk 7 above)
Comment (ArielG-NV) [reply tools/render-test/shader-input-layout.cpp]: the code assumes the type is 0 or 1, anything in-between does not matter

the code during initialization only has: one, zero, one & zero, as the initialization parameters, as a result the code works for the purposes of testing

I will comment as such

**Comment by ArielG-NV** [comment]
I found some things to change*, I will adjust them quickly today before more review

File: tools/gfx-unit-test/format-unit-tests.cpp (hunk 8 above)
Comment (jkwak-work) [reply tools/gfx-unit-test/format-unit-tests.cpp]: If I were you, I would remove it because it is not hard to bring it back when we can actually test it.

If you like to keep it as it is, you will need to leave a comment and describe what needs to be done.

File: source/slang/glsl.meta.slang (hunk 2 above)
Comment (jkwak-work) [reply source/slang/glsl.meta.slang]: Are you going to fix the typo?

File: source/slang/slang-check-decl.cpp (hunk 3 above)
Comment (jkwak-work) [reply source/slang/slang-check-decl.cpp:962]: The document described the case where both readonly and writeonly are used.
Can we have a test case for this case? I guess we can test if it can use "imageSize" and "length".
```
A variable could be qualified as both readonly and writeonly, disallowing both read and write.
Such variables can still be used with some queries, for example imageSize() and .length().
```

File: source/slang/slang-check-modifier.cpp [hunk 9]
```diff
@@ -1092,15 +1094,26 @@ namespace Slang
...
+        case ASTNodeType::GLSLReadOnlyModifier:
+        case ASTNodeType::GLSLVolatileModifier:
+        case ASTNodeType::GLSLRestrictModifier:
+        {
+            return (as<VarDeclBase>(decl) && (isGlobalDecl(decl)) || as<ParamDecl>(decl) || as<GLSLInterfaceBlockDecl>(decl))
+                || as<StructDecl>(getParentDecl(decl)) && isGLSLInput;
+        }
         case ASTNodeType::GloballyCoherentModifier:
+        {
+            return (isGlobalDecl(decl) && ((as<VarDeclBase>(decl) && isGLSLInput) || as<VarDecl>(decl))) || as<ParamDecl>(decl) && isGLSLInput
```
Comment (jkwak-work) [inline source/slang/slang-check-modifier.cpp]: This line is very hard to read.
And I think you are changing the behavior on HLSL side.
Before your change, as<VarDecl>(decl) had to be true to return true. But after your change it doesn't need to be.

When isGLSLInput is false, it became,
```
return (isGlobalDecl && VarDecl) || struct || interface;
```
That is different from
```
return VarDecl && (isGlobalDecl || struct || interface);
```

I will be also sensetive about the order of the evaluation.
It seems that "isGlobalDecl" is called always after "as<VarDeclBase>(decl)" or "as<VarDecl(decl)" is true; not before.

You can separate GLSL from HLSL like following:
```
if (isGLSLInput)
    return ((VarDeclBase || VarDecl) && isGlobalDecl) || ParamDecl || struct || interface;
return as<VarDecl>(decl) && (isGlobalDecl(decl) || as<StructDecl>(getParentDecl(decl)) || as<GLSLInterfaceBlockDecl>(decl));
```

OpenGL document shows two examples.
Can we have them in our test?
```
coherent buffer Block {
  readonly vec4 member1;
  vec4 member2;
};
buffer Block {
  coherent readonly vec4 member1;
  coherent vec4 member2;
};
```

File: source/slang/slang-diagnostic-defs.h [hunk 10]
```diff
@@ -354,6 +356,8 @@ DIAGNOSTIC(30098, Error, nonStaticMemberFunctionNotAllowedAsDiffOperand, "non-st
 
 DIAGNOSTIC(30099, Error, sizeOfArgumentIsInvalid, "argument to sizeof is invalid")
 
+DIAGNOSTIC(30101, Error, readingFromWriteOnly, "right of '=' is writeonly (cannot be used for reading). Check modifiers.")
```
Comment (jkwak-work) [inline source/slang/slang-diagnostic-defs.h]: I think it will be better to be a little more generic like "Cannot read from writeonly variable."
I don't think reading operation is limited to "operator=", because the initialization expression could be in other forms.

File: source/slang/slang-emit-spirv.cpp [hunk 11]
```diff
@@ -1895,6 +1926,7 @@ struct SPIRVEmitContext
         //
         // SPIR-V requires that the sampled/rw info on the image isn't unknown
         SLANG_ASSERT(sampled == sampledImage || sampled == readWriteImage);
+        if(isMultisampled) requireSPIRVCapability(SpvCapabilityStorageImageMultisample);
```
Comment (jkwak-work) [inline source/slang/slang-emit-spirv.cpp]: Can you make this two lines?
It is difficult to debug when they are all in one line.

File: source/slang/slang-emit-spirv.cpp [hunk 12]
```diff
@@ -1765,6 +1765,37 @@ struct SPIRVEmitContext
...
+            if(setCapabilityMask == SpvCapabilityShader) return;
+            requireSPIRVCapability(SpvCapabilityShader);
+        case SpvImageFormatR64ui:
+        case SpvImageFormatR64i:
+            if(setCapabilityMask == SpvCapabilityInt64ImageEXT) return;
+            ensureExtensionDeclaration(UnownedStringSlice("SPV_EXT_shader_image_int64"));
+            requireSPIRVCapability(SpvCapabilityInt64ImageEXT);
+        default:
+            if(setCapabilityMask == SpvCapabilityStorageImageExtendedFormats) return;
+            requireSPIRVCapability(SpvCapabilityStorageImageExtendedFormats);
```
Comment (jkwak-work) [inline source/slang/slang-emit-spirv.cpp]: Is this intentionally fall through?
If yes, can you use [[fallthrough]];

File: source/slang/slang-ir-inst-defs.h [hunk 13]
```diff
@@ -155,6 +150,11 @@ INST(Nop, nop, 0, 0)
             /* TextureTypeBase */
                 /* TextureType */
                 INST(TextureType, TextureType, 8, HOISTABLE)
+                INST(TextureShape1DType, TextureShape1DType, 0, HOISTABLE)
+                INST(TextureShape2DType, TextureShape1DType, 0, HOISTABLE)
+                INST(TextureShape3DType, TextureShape1DType, 0, HOISTABLE)
+                INST(TextureShapeCubeType, TextureShape1DType, 0, HOISTABLE)
+                INST(TextureShapeBufferType, TextureShapeBufferType, 0, HOISTABLE)
```
Comment (jkwak-work) [inline source/slang/slang-ir-inst-defs.h]: Why did they become a part of TextureTypeBase?

**Review by jkwak-work — COMMENTED** [review]
I left some comments.
I haven't read testing code yet.

File: source/slang/slang-check-modifier.cpp (hunk 9 above)
Comment (ArielG-NV) [reply source/slang/slang-check-modifier.cpp]: will adjust the decl values since they are done confusingly/wrong

but I will need to adjust it beyond what was suggested

AllowGLSL supports HLSL syntax and glsl syntax

File: source/slang/slang-diagnostic-defs.h (hunk 10 above)
Comment (ArielG-NV) [reply source/slang/slang-diagnostic-defs.h]: I will make it more generic, I agree with the usage.

File: source/slang/slang-ir-inst-defs.h (hunk 13 above)
Comment (ArielG-NV) [reply source/slang/slang-ir-inst-defs.h]: likely don't need the change I made now regardless

File: tools/gfx-unit-test/format-unit-tests.cpp (hunk 8 above)
Comment (ArielG-NV) [reply tools/gfx-unit-test/format-unit-tests.cpp]: I will remove the code, the comment would likely rot as we wait for hardware to eventually support non storage 64bit (which is an if, rather than a when).

File: tests/glsl-intrinsic/image/i64imageDim.slang [hunk 14]
```diff
@@ -0,0 +1,483 @@
...
+// As per vulkan specification: https://registry.khronos.org/vulkan/specs/1.3/html/chap34.html#formats-definition
+// i64/u64 images do not require sampled image support; support is currently unlikley on hardware
+//#define test_when_hardware_supports_i64_and_u64_sampled_textures
+// i64/u64 images do not require texel buffers support; support is currently unlikley on hardware
+//#define TEST_when_hardware_supports_i64_and_u64_buffers
+
+//TEST_INPUT:ubuffer(data=[0], stride=4):out,name=outputBuffer
+buffer MyBlockName2
+{
+    uint data[];
```
Comment (jkwak-work) [inline tests/glsl-intrinsic/image/i64imageDim.slang]: Shouldn't this have a size?
```
uint data[1];
```

File: tests/glsl-intrinsic/image/i64imageDim.slang [hunk 15]
```diff
@@ -0,0 +1,483 @@
...
+} outputBuffer;
+
+layout(local_size_x = 1) in;
+
+//TEST_INPUT: RWTexture1D(format=R64_SINT, size=4, content=one, mipMaps = 1):name image_1d
```
Comment (jkwak-work) [inline tests/glsl-intrinsic/image/i64imageDim.slang:20]: Aren't you missing "=" between "name" and "image_1d"?

File: tests/glsl-intrinsic/image/i64imageDim.slang [hunk 16]
```diff
@@ -0,0 +1,483 @@
...
+#ifdef test_when_hardware_supports_i64_and_u64_sampled_textures
+        && imageSize(image_2dMultiSample) == ivec2(4)
+        && imageSize(image_2dMultiSampleArray) == ivec3(4, 4, 2)
+#endif // #ifdef test_when_hardware_supports_i64_and_u64_sampled_textures
+        ;
+}
+bool checkAllImageLoad()
+{
+    return true
+        && imageLoad(image_1d, 0).x == i64vec4(1).x
```
Comment (jkwak-work) [inline tests/glsl-intrinsic/image/i64imageDim.slang]: Why do you need to use i64vec4(1).x?
Can it be just 1LL?

File: tests/glsl-intrinsic/image/iimageDim.slang [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+    loadCheck = loadCheck && load_image_3d() == 0;
+    imageStore(image_cube, ivec3(0), ivec4(0));
+    loadCheck = loadCheck && load_image_cube() == 0;
+    imageStore(image_cubeArray, ivec3(0), ivec4(0));
+    loadCheck = loadCheck && load_image_cubeArray() == 0;
+    imageStore(image_2dMultiSampleArray, ivec3(0), 1, ivec4(0));
+    loadCheck = loadCheck && load_image_2dMultiSampleArray() == 0;
+    resetAllImageValues();
+    return loadCheck;
+        ;
```
Comment (jkwak-work) [inline tests/glsl-intrinsic/image/iimageDim.slang]: Probably not needed line

File: tests/glsl-intrinsic/image/iimageDim.slang [hunk 18]
```diff
@@ -0,0 +1,418 @@
...
+        && checkAllImageAtomicCompSwap()
+        ;
+}
+
+// CHECK_GLSL: void main(
+// CHECK_SPV: OpEntryPoint
+void computeMain()
+{
+    outputBuffer.data[0] = true
+        && imageOperations()
```
Comment (jkwak-work) [inline tests/glsl-intrinsic/image/iimageDim.slang]: I think you can just inline the function "imageOperations" here.
```
void computeMain()
{
outputBuffer.data[0] =true
        && checkAllImageSize()
        && checkAllImageLoad()
        && checkAllImageStore()
        && checkAllImageSamples()
        && checkAllImageAtomicAdd()
        && checkAllImageAtomicExchange()
        && checkAllImageAtomicMin()
        && checkAllImageAtomicMax()
        && checkAllImageAtomicAnd()
        && checkAllImageAtomicOr()
        && checkAllImageAtomicXor()
        && checkAllImageAtomicCompSwap()
        ;
}

File: tests/glsl/ssboMemberMemoryQualifier.slang [hunk 19]
```diff
@@ -0,0 +1,48 @@
+//DIAGNOSTIC_TEST:SIMPLE(filecheck=CHECK-GLSL): -stage compute -entry computeMain -allow-glsl -target glsl
+//DIAGNOSTIC_TEST:SIMPLE(filecheck=CHECK-SPV): -stage compute -entry computeMain -allow-glsl -target spirv -emit-spirv-directly
```
Comment (jkwak-work) [inline tests/glsl/ssboMemberMemoryQualifier.slang]: What is DIAGNOSTIC_TEST?
And can we have a runtime test like this?
```
//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl
//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
```
I don't know if DIAGNOSTIC_TEST does it.

File: tools/gfx/vulkan/vk-device.cpp [hunk 20]
```diff
@@ -1638,6 +1650,96 @@ Result DeviceImpl::createTextureResource(
...
+            gfxGetFormatInfo(desc.format, &formatInfo);
+            uint32_t data = 0;
+            VkClearColorValue clearColor;
+            switch(formatInfo.channelType)
+            {
+            case SLANG_SCALAR_TYPE_INT32:
+                memcpy(&data, initData->data, sizeof(data));
+                break;
+            case SLANG_SCALAR_TYPE_UINT32:
+                memcpy(&data, initData->data, sizeof(data));
```
Comment (jkwak-work) [inline tools/gfx/vulkan/vk-device.cpp]: Maybe missing a "break"

File: tools/gfx/vulkan/vk-device.cpp [hunk 21]
```diff
@@ -1638,6 +1650,96 @@ Result DeviceImpl::createTextureResource(
...
+            case SLANG_SCALAR_TYPE_INT64:
+            {
+                int32_t tmpData = int32_t(((int64_t*)initData->data)[0]);
```
Comment (jkwak-work) [inline tools/gfx/vulkan/vk-device.cpp]: I would prefer "reinterpret_cast" over c-style type cast.

File: tools/gfx/vulkan/vk-device.cpp [hunk 22]
```diff
@@ -1638,6 +1650,96 @@ Result DeviceImpl::createTextureResource(
...
+            case SLANG_SCALAR_TYPE_FLOAT16:
+            {
+                float tmpData = HalfToFloat(((uint16_t*)initData->data)[0]);
+                memcpy(&data, &tmpData, sizeof(data));
+                break;
+            }
+            case SLANG_SCALAR_TYPE_FLOAT32:
+            {
+                float tmpData = float(((float*)initData->data)[0]);
+                memcpy(&data, &tmpData, sizeof(data));
```
Comment (jkwak-work) [inline tools/gfx/vulkan/vk-device.cpp]: I think it will be easier if you just assign the values without going through "data".
```
for (int i =0; i < 4; ++i)
clearColor.float32[i] = *reinterpret_cast<float*>(initData->data);
```

File: tools/gfx/vulkan/vk-device.cpp [hunk 23]
```diff
@@ -2154,6 +2256,19 @@ Result DeviceImpl::createBufferView(
...
                 info.offset = offset;
                 info.range = size;
+                VkBufferUsageFlags2CreateInfoKHR bufferViewUsage;
+                bufferViewUsage = {};
+                bufferViewUsage.sType = VK_STRUCTURE_TYPE_BUFFER_USAGE_FLAGS_2_CREATE_INFO_KHR;
+                info.pNext = &bufferViewUsage;
+                if (desc.type == IResourceView::Type::UnorderedAccess)
+                {
+                    bufferViewUsage.usage = VK_BUFFER_USAGE_2_STORAGE_TEXEL_BUFFER_BIT_KHR;
+                    info.pNext = &bufferViewUsage;
```
Comment (jkwak-work) [inline tools/gfx/vulkan/vk-device.cpp]: It doesn't look like you need this line.

File: tools/render-test/shader-input-layout.cpp (hunk 7 above)
Comment (jkwak-work) [reply tools/render-test/shader-input-layout.cpp]: If the values in srcPixels are expected to be zero or one, why do we need to divide them with 255?
And is the input values never going to be changed to values other than zero and one?

The intention here looks to be converting the image format from RGBA8 to R64_UINT, as an example.
I think you need to do a proper conversion.

**Review by jkwak-work — COMMENTED** [review]
I left more comments.

File: tests/glsl-intrinsic/image/i64imageDim.slang (hunk 14 above)
Comment (ArielG-NV) [reply tests/glsl-intrinsic/image/i64imageDim.slang]: for future reference: unbound arrays are legal in glsl if final param (this is preferred to make testing modifications easier): https://www.khronos.org/opengl/wiki/Shader_Storage_Buffer_Object

I can add the 1 though

File: source/slang/slang-ir-inst-defs.h (hunk 13 above)
Comment (jkwak-work) [reply source/slang/slang-ir-inst-defs.h]: I don't know what you mean by that.
That sounds like how Yoda would say.
If the change is not needed, please don't change it.

File: tests/glsl-intrinsic/image/i64imageDim.slang (hunk 16 above)
Comment (ArielG-NV) [reply tests/glsl-intrinsic/image/i64imageDim.slang]: I can simplify these (the format was kept consistent across tests for making changing easier, but removing it is completely fine)

File: tests/glsl-intrinsic/image/i64imageDim.slang (hunk 15 above)
Comment (ArielG-NV) [reply tests/glsl-intrinsic/image/i64imageDim.slang:20]: ironically, we do not (the code below is from the parser, note `// Optionally consume '='`):
```
                    else if (parser.AdvanceIf("name"))
                    {
                        // Optionally consume '=' 
                        if (parser.NextToken().Type == Misc::TokenType::OpAssign)
                        {
                            parser.ReadToken();
                        }

                        ioField.name = parseName(parser);
                    }
```

File: tests/glsl/ssboMemberMemoryQualifier.slang (hunk 19 above)
Comment (ArielG-NV) [reply tests/glsl/ssboMemberMemoryQualifier.slang]: DIAGNOSTIC_TEST is just syntax candy. I was using it since I was basing this file off of the error check files I made which were based on a file that use DIAGNOSTIC_TEST ('unknown-opcode.slang` is an example of this).

We can have a runtime test; we don't really need it though since the goal is "make sure we can add qualifiers here" as an opposite to "make sure we cannot add qualifier here" tests. Testing if we can write 1 to a int is not really helpful since the regression is all "compiler enforcing things" and decorations which indirectly effect runtime. If glslang does not compile glslang would complain, if glslang compiles, glslang won't complain.

Comment (jkwak-work) [reply tests/glsl/ssboMemberMemoryQualifier.slang]: I still don't know what DIAGNSTIC_TEST is.
Can you tell me why this test uses DIAGNSTIC_TEST not TEST?

If the test is not meant for runtime result, we don't need to test it here. That is fine.

FYI, the other day, Yong told me that we cannot rely on the error messages on the downstream compilers. Slang must catch and report the syntactic errors without relying on the error messages from glslang.

Comment (ArielG-NV) [reply tests/glsl/ssboMemberMemoryQualifier.slang]: 1. "Can you tell me why this test uses DIAGNSTIC_TEST not TEST?" because the code I based this code off of used DIAGNSTIC_TEST, they function identical, I will switch to TEST, there is no issue in doing so 

2.👍 

3. we don't rely on glslang error messages for any of the following code, I mean it as in: this is all compile time syntax candy for as far as the programmer will experience, the compiler (Slang) will error if you break syntax rules (and I am checking for these)

File: tests/glsl-intrinsic/image/i64imageDim.slang (hunk 16 above)
Comment (ArielG-NV) [reply tests/glsl-intrinsic/image/i64imageDim.slang]: I checked; 1 casts completely fine as well

File: tools/render-test/shader-input-layout.cpp (hunk 7 above)
Comment (ArielG-NV) [reply tools/render-test/shader-input-layout.cpp]: 0 or 1 was a unorm description of what it did under the-hood, I explained the process with unorm because the data is actually stored as uint8_t, 1/255*255 gives a float representation (for casting) which is how floatToHalf works. The existing system was only designed to handle 0 or 1, any other value and you are at the mercy of IEEE and floating point inaccuracies

Comment (ArielG-NV) [reply tools/render-test/shader-input-layout.cpp]: thinking about a proper conversion:

Comment (ArielG-NV) [reply tools/render-test/shader-input-layout.cpp]: lets say our input system says: load zeroPointFive, this would require the code before this which is loading the initial data to setup 128,

lets say we input two, this would require the code before this which is loading the initial data to setup 2 since 512 is not an option

Comment (jkwak-work) [reply tools/render-test/shader-input-layout.cpp]: I don't understand what you are saying.
Here is how I see it.
When we convert from RGBA8 format to RGBA32F, we do the following conversion.
```
rgba32f[0] = float(rgba8[0]) / 255.f;
```
When we convert from RGBA8 to RGBA16, as an example, we do the following,
```
rgba16[0] = rgba8[0] << 8;
```
Your implementation works for only one case where the input is 8bit integer and output is a floating point. It wouldn't work when the output is an integer format.

I don't know what you mean by 1.
It could mean a near dark color in RGBA8 format. (1/255).
Or it could mean a full white color in RGBA8 format (255/255).
Your expression is very hard to understand even with the examples.

File: tests/glsl/ssboMemberMemoryQualifier.slang (hunk 19 above)
Comment (jkwak-work) [reply tests/glsl/ssboMemberMemoryQualifier.slang]: It looks like DIAGNOSTIC_TEST is meant for testing error cases.
```
struct TestOptions
{
    enum Type
    {
        Normal,             ///< A regular test
        Diagnostic,         ///< Diagnostic tests will always run (as form of failure is being tested)
    };
```
So if your intention is to testing failures or error messages, it seems appropriate to use it.

File: tools/render-test/shader-input-layout.cpp (hunk 7 above)
Comment (ArielG-NV) [reply tools/render-test/shader-input-layout.cpp]: my implementation worked for all types as long as it was 0 or 1 as input. It simply took a float which was expected to be 0.0f or 1.0f, and casted it to another type.

I changed the implementation (as per the request) to work with non 0/1 inputs as well for the future.

about the new implementation: I did not want to make code that did something in the nature of: `rgba16[0] = rgba8[0] << 8;` since then when adding all 12 types we would have a big bundle of code which would be hard to iterate on, change, and reuse. I wanted to keep everything easy to understand and modify to allow easy additions of functionality. I think the new one is quite easy to work around.

File: source/slang/glsl.meta.slang [hunk 24]
```diff
@@ -2821,6 +2821,453 @@ public vec4 shadow2DProjLod(sampler2DShadow sampler, vec4 coord, float lod)
...
+    struct ImageTypeInfo
+    {
+        const char *prefix;
+        const char *type;
+        bool isInt;
+        const char *SPVTypePrefix;
+        const char *SPVSubTypePrefix;
+        const char *SPVAtomicSuffix;
+    }
+    static const kTypeWithPrefix[] =
```
Comment (jkwak-work) [inline source/slang/glsl.meta.slang]: Please change this to be more explicit.
```
};

static const ImageTypeInfo kTypeWithPrefix[] = 
```

File: source/slang/slang-check-modifier.cpp [hunk 25]
```diff
@@ -1092,17 +1094,25 @@ namespace Slang
...
+            return (as<VarDeclBase>(decl) && (isGlobalDecl(decl)) || as<ParamDecl>(decl) || as<GLSLInterfaceBlockDecl>(decl))
+                // If allowing GLSL syntax, also allow the following
+                || isGLSLInput && (as<StructDecl>(getParentDecl(decl)) && isGlobalDecl(getParentDecl(decl)));
         case ASTNodeType::GloballyCoherentModifier:
         case ASTNodeType::HLSLVolatileModifier:
-            return as<VarDecl>(decl) && (isGlobalDecl(decl) || as<StructDecl>(getParentDecl(decl)) || as<GLSLInterfaceBlockDecl>(decl));
+            return as<VarDecl>(decl) && (isGlobalDecl(decl) || as<StructDecl>(getParentDecl(decl)) || as<GLSLInterfaceBlockDecl>(decl))
+                // If allowing GLSL syntax, also allow the following
+                || isGLSLInput && (as<VarDeclBase>(decl) && isGlobalDecl(decl) || as<ParamDecl>(decl) || (as<StructDecl>(getParentDecl(decl)) && isGlobalDecl(getParentDecl(decl))))
+                ;
```
Comment (jkwak-work) [inline source/slang/slang-check-modifier.cpp]: Please split "isGLSLInput" into a new if-statement as I suggested.
This is still hard to read, which is easier to make a mistake.

File: tools/gfx/vulkan/vk-device.cpp [hunk 26]
```diff
@@ -2154,6 +2257,20 @@ Result DeviceImpl::createBufferView(
...
+                if (desc.type == IResourceView::Type::UnorderedAccess)
+                {
+                    bufferViewUsage.usage = VK_BUFFER_USAGE_2_STORAGE_TEXEL_BUFFER_BIT_KHR;
+                }
+                else if (desc.type == IResourceView::Type::ShaderResource)
+                {
+                    bufferViewUsage.usage = VK_BUFFER_USAGE_2_UNIFORM_TEXEL_BUFFER_BIT_KHR;
+                }
+                else
+                    assert(!"unhandled");
```
Comment (jkwak-work) [inline tools/gfx/vulkan/vk-device.cpp]: Please wrap the line with "{" and "}".
Depending on how "assert" macro is defined, this line could disappear and unwanted next line may become a part of "else".

File: tools/render-test/shader-input-layout.cpp [hunk 27]
```diff
@@ -1100,7 +1126,109 @@ namespace renderer_test
...
+    {
+        // well define int8_t = uint8_t
+        if constexpr (std::is_same<int8_t, T>())
+        {
+            if (in > 127)
+            {
+                out = T(127);
+                return;
+            }
+        }
```
Comment (jkwak-work) [inline tools/render-test/shader-input-layout.cpp]: Why is only one type treated specially?

**Review by jkwak-work — COMMENTED** [review]
I left a few comments.
We are almost getting there.

File: tools/gfx/vulkan/vk-device.cpp (hunk 26 above)
Comment (ArielG-NV) [reply tools/gfx/vulkan/vk-device.cpp]: I had no idea assert is a macro sometimes, thank you for sharing~ I 100% should NOT be doing this with assert then...

File: tools/render-test/shader-input-layout.cpp (hunk 27 above)
Comment (ArielG-NV) [reply tools/render-test/shader-input-layout.cpp]: uint8_t is the src type, max size is 255.

int16, uint16, half all can safely store the type lower and upper bounds in normal c++ rules with an explicit cast or with logic (FloatToHalf).

int8_t is the only type which cannot cast 255 into itself within a strictly positive set of numbers, the spec states that bit pattern is preserved, so this would end-up being [0,127],[-1,-128]. Since testing code does not assign negatives I decided to follow this pattern and just keep things consistent with a positive only int.

Comment (ArielG-NV) [reply tools/render-test/shader-input-layout.cpp]: I can add a way to support negative fairly trivially if you prefer and believe it is the correct method
